### ISSUE TITLE
offer a standard scripting interface for package / lifecycle scripts

### ISSUE THREAD

**Issue by othiym23** [npm/npm#11291]
This has been discussed in many places, but I'm not sure that it's been filed as a feature request now: the current cross-platform behavior requires people writing package scripts to use a very tightly-constrained subset of shell scripting syntax so that both Windows and Unix users can use the same scripts. This basically would involve choosing a simple subset of shell command syntax and then wrapping that around something like `shelljs`. @TheAlphaNerd had some thoughts as well.

See also #1543 (add a `npm exec`), which is closely related.

**Comment by boneskull** [comment]
idea 1

1.  Include `shelljs` as dependency

2.  Expose it as command `shjs`

3.  Use it like so:

  ```json
  {
    "scripts": {
      "build": "npm shjs path/to/build/module.js"
    }
  }
  ```

4. Potential improvement: by default, look in `.scripts/` dir local to project for scripts.

**Comment by boneskull** [comment]
idea 2

1. Keep idea 1
2. Write a wrapper around `shjs` which accepts options
3. Each option corresponds to a `shelljs` command.  Arguments to the command could be delimited, or we could use something like subargs (think using transforms w/ browserify on command-line).
4.  Would look like this:

  ```json
  {
    "scripts": {
      "clean": "npm shjs --rm [ -rf build_dir ]"
    }
  }
  ```
5.  I'm not a huge subargs fan so if anyone has a better idea, please propose

**Comment by boneskull** [comment]
(instead of command being `shjs`, maybe `shell` is fine.)

**Comment by boneskull** [comment]
I'm not sure if the shelljs team would be interested in a PR to accept options in this manner.  If they are, this implementation should take place in shelljs.

For the implementation, the quickest route from A -> B here might just be dumping a `.js` file in a temp dir and pointing `shjs` at it.

cc @ariporad @nfischer

**Comment by nfischer** [comment]
@boneskull It sounds like your idea 2 would be along the lines of shelljs/shelljs#244. Correct me if I'm mistaken.

**Comment by ariporad** [comment]
@boneskull: Agreed, this sounds like shelljs/shelljs#244.

We're hoping to ship that as a separate module ('shx') at the same time as v0.7.0.

**Comment by dthree** [comment]
I have a project releasing in a few days which may fit this role perfectly. That's up to you guys obviously. I'll fill you in as soon as I publicly release it.

**Comment by nfischer** [comment]
For reference, we've now started the initial repo, and just need to flesh out more of the implementation. Expect an initial release in a week or two.

https://github.com/shelljs/shx

**Comment by dthree** [comment]
@othiym23 et al,

Hey. I know I'm late on the draw now, however I just released a library two days ago that I've been working on for a few months. I ran into this issue and thought that it might just be an excellent fit to solve this problem.

https://github.com/dthree/cash

Again, it's new, but it's very rapidly picking up steam. I love ShellJS (and I promote it in Cash's README), however on this particular issue there's a few things Cash seems to be a really good fit for:

- Cash was built around full implementations of commands, aimed at POSIX-compliance. While it's still on its way there, it supports many more options for commands and so gives better cross-platform bash "feel".

- It supports complex interpretations such as command piping. My concept is that one could simply literally interpret the NPM script commands through Cash (without explicitly calling Cash) so it is completely transparent - the user would just assume they are writing in bash. As it aims for POSIX-compliance, there doesn't have to be some separate syntax manual for what is or isn't supported in NPM scripts.

Is this something we could look at? Thanks!

**Comment by nfischer** [comment]
Cash looks pretty cool! I haven't tried it out yet, but it definitely seems like a comprehensive solution.

I think `shx` will be a nice solution for users who want to be able to rely on ShellJS semantics. Cash is probably a good solution for people who don't like those semantics, or would like features shelljs doesn't offer. I think either of these would be good solutions.

Also, :+1: for implementing pipes (ShellJS will release this feature in v07, but it was lacking it for many years).

**Comment by dthree** [comment]
Thanks! And :+1: for ShellJS, it's really awesome!

**Comment by sindresorhus** [comment]
:+1: Cash is really good. It would be amazing to have it as a part of npm, enabling writing cross-platform run scripts with what feels like plain unix commands.

**Comment by iarna** [comment]
If as you say, it achieves POSIX compatibility then I would expect we would be very keen to integrate it. And yes, passing them through something like Cash is exactly what we wanted to do. Obviously that's a long way off right now, and it's a big project, but I would love to see it happen.

While the command support is important, it's actually the shell command line parsing that I think is the biggest challenge here. Env variable substitution, command substitution, a variety of quoting styles, inline math, globbing, etc. If a project can nail that down, writing POSIX-compatible implementations of the core shell tools should be pretty easy (in large part because that's so easy to split work up on).

**Comment by boneskull** [comment]
Indeed, I saw Cash and thought it'd be perfect for this.

**Comment by dthree** [comment]
@iarna I agree. Cash depends on [Vorpal](https://github.com/dthree/vorpal) and I do intend to mature it even further in the regards you lay out. I also intend to give Cash full support of bash basics. It's definitely got a lot already, so I'll keep at it!

Perhaps at some point I can put together a demo fork of NPM using Cash.

---

@boneskull thanks!

**Comment by ariporad** [comment]
Agreed. Despite the fact that it might hurt shelljs's download numbers (:cry:), I think that cash is really awesome, and it would be better to just be able to write bash on every platform, and not worry about it.

**Comment by dthree** [comment]
:cry: Hurting ShellJS is that last thing I want to do, which is part of why I promote it in my project. I don't really think the two conflict - they're just tailored for different purposes. ShellJS excels by far in a programmatic environment.

**Comment by ariporad** [comment]
I know, it was mostly a joke (Sorry, I forgot that sarcasm doesn't work
well over the internet). But regardless, thanks for making something as
awesome as cash.


Ari

**Comment by nfischer** [comment]
I don't necessarily think it's "better" to write bash on every platform. Bash is nice, and it's nice to have that feature. I like shelljs because it uses JavaScript syntax, which I feel is easier to use than Bash's (sometimes strange) syntax. Writing simple commands is much easier in bash, and I think that's the strength of cash. I think writing scripts in shelljs is sometimes more robust than bash, however, and that's my aim for the shelljs project.

I'd like to move the project toward more sensible defaults, ex. Making iterating over lines of output trivial, instead of forcing the programmer to jump through the hoops to not iterate over each space delimited word. Posix defaults are the right move for a shell like cash, but I think it's clear that shelljs can move on to something nicer for the programmer now that cash is covering that. Shelljs will always have posix compatibility as a priority, but now we'll probably try to only do it where it's an advantageous feature.

**Comment by iarna** [comment]
I don't really see the two as competing. For npm's case (to bring this back on topic), we need something that can evaluate POSIX shell (at minimum) command lines, as that's what lifecycle scripts have always been. This isn't a value judgement, just reproducing what `process.exec` was doing in a platform independent manor.

**Comment by ariporad** [comment]
@iarna: I vote for cash (Sorry for the diversion, my bad), although it might be better to wait for `cash@1.0.0`.

**Comment by nfischer** [comment]
I don't think shelljs will every be designed to parse bash code and execute it, so it (in itself) isn't quite right for that as part of npm. Shx (our new project) is slightly closer in appearance, but still isn't a true shell. Cash sounds like it aims to parse bash syntax, so it (or something like it) is probably a better option for npm here.
